proxy/coco/src/request.rs

```diff
@@ -0,0 +1,311 @@
+        Request {
+            urn: self.urn,
+            attempts: Attempts {
+                queries: self.attempts.queries + 1,
+                ..self.attempts
+            },
+            timestamp,
+            state: PhantomData,
+        }
+    }
+}
+
+impl<T> Request<IsRequested, T> {
+    /// Transition the `Request` from the `IsRequested` state to the `Found` state.
+    ///
+    /// This signifies that the `Request` found its first peer and will be ready to attempt to
+    /// clone from the peer.
+    pub fn first_peer(self, peer: PeerId, timestamp: T) -> Request<Found, T> {
+        let mut peers = HashMap::new();
+        peers.insert(peer, Status::Available);
+        Request {
+            urn: self.urn,
+            attempts: self.attempts,
+            timestamp,
+            state: Found { peers },
+        }
+    }
+}
+
+impl<T> Request<Found, T> {
+    /// Transition the `Request` from the `Found` state to the `Cloning` state.
+    ///
+    /// This signifies that the `Request` is attempting to clone from the provided `peer`.
+    pub fn cloning(
+        self,
+        max_queries: Queries,
+        max_clones: Clones,
+        peer: PeerId,
+        timestamp: T,
+    ) -> Either<Request<TimedOut, T>, Request<Cloning, T>>
+    where
+        T: Clone,
+    {
+        let mut peers = self.state.peers;
+        peers
+            .entry(peer)
+            .and_modify(|status| *status = Status::InProgress)
+            .or_insert(Status::InProgress);
+        let this = Request {
+            urn: self.urn,
+            attempts: Attempts {
+                queries: self.attempts.queries,
+                clones: self.attempts.clones + 1,
+            },
+            timestamp: timestamp.clone(),
+            state: Cloning { peers },
+        };
+        this.timed_out(max_queries, max_clones, timestamp).flip()
+    }
+
+    /// Transition the `Request` from the `Found` back to the `IsRequested` state.
+    ///
+    /// This signifies that the `Request` has exhausted its list of peers to attempt cloning from
+    /// and needs to re-attempt the request for the search.
+    ///
+    /// Should be used in tandem with [`HasPeers::all_failed`] to ensure that we transition back
+    /// when all peers have failed to clone.
+    #[allow(clippy::missing_const_for_fn)]
+    pub fn failed(self, timestamp: T) -> Request<IsRequested, T> {
+        Request {
+            urn: self.urn,
+            attempts: self.attempts,
+            timestamp,
+            state: PhantomData,
+        }
+    }
+}
+
+impl<T> Request<Cloning, T> {
+    /// Transition from the `Cloning` state back to the `Found` state.
+    ///
+    /// This signifies that the `peer` failed to clone the identity and we mark it as failed.
+    pub fn failed(self, peer: PeerId, timestamp: T) -> Request<Found, T> {
+        let mut peers = self.state.peers;
+        // TODO(finto): It's weird if it didn't exist but buh
```

Comment:
😂  How does one resolve this `TODO`?

Reply:
Probably with `dd` in vim :shushing_face: